### PR TITLE
BUG: fix overflow in stats.boxcox_normmax when method='mle'

### PR DESCRIPTION
**Pull request by xuefeng-xu**
#### Reference issue

#18389 (overflow also happens in `boxcox`)

#### What does this implement/fix?

Code is inspired by https://github.com/scipy/scipy/pull/18852#issuecomment-1657858886 and #19016
1. Fix overflow in `stats.boxcox_normmax` when method='mle'
2. Adjust the output lambda to ensure that the transformed data is not infinite

#### Additional information

The Box-Cox transformation is $\phi(x)=(x^\lambda - 1)/\lambda$ for $\lambda\neq0$

Modify the variance formula since they don't affect the result
1. Remove the offset $1/\lambda$ #10072
2. Change the denominator to $|\lambda|$ (compute log)

Therefore,
$\text{Var}((x^\lambda - 1)/\lambda)=\text{Var}(x^\lambda /\lambda)=\text{Var}(x^\lambda /|\lambda|)$

Then, compute the log
$\log(x^\lambda /|\lambda|)=\lambda\log x - \log|\lambda|$

Finally, compute the log of the variance of transformed data using `_log_var`

~~To adjust the final $\lambda$, ensure $x^\lambda$ does not cause overflow, which is much simpler than the inverse of $\phi(x)$~~
~~So $\lambda^*=\log c / \log x$, $c$ is the maximum of floating point.~~

Use the exact approach to adjust the final lambda.
https://www.wolframalpha.com/input?i=solve+%28x%5El+-1%29+%2F+l+%3D+y+for+l

### CODE REVIEW

File: scipy/stats/_morestats.py [hunk 1]
```diff
@@ -1288,6 +1304,28 @@ def _all(x):
...
+    else:
+        # Test if the optimal lambda causes overflow
+        max_x = np.max(x, axis=0)
+        istransinf = np.isinf(special.boxcox(max_x, res))
+        dtype = x.dtype if np.issubdtype(x.dtype, np.floating) else np.float64
+        if np.any(istransinf):
+            warnings.warn(
+                f"The optimal lambda is {res}, but the returned lambda is"
+                " the constrained optimum to ensure that maximum of"
+                f" transformed data does not cause overflow in {dtype}."
```
Comment (lucascolley) [inline scipy/stats/_morestats.py:1316]:
```suggestion
                f"The optimal lambda is {res}, but the returned lambda is "
                f"the constrained optimum to ensure that maximum of " 
                f"transformed data does not cause overflow in {dtype}."
```

File: scipy/stats/_morestats.py [hunk 2]
```diff
@@ -1274,6 +1289,7 @@ def _all(x):
 
     optimfunc = methods[method]
 
+    x = np.asarray(x)
```
Comment (lucascolley) [inline scipy/stats/_morestats.py:1292]: is there a reason for introducing this here rather than in the new `else` branch below?

Comment (xuefeng-xu) [reply scipy/stats/_morestats.py:1292]: No, I will move it to the else branch.

File: scipy/stats/_morestats.py (hunk 1 above)
Comment (lucascolley) [reply scipy/stats/_morestats.py:1316]: We don't, it is purely a style preference for me. If you think it looks better without the `f`, feel free to not include it, but I think it looks better to have the lines aligned.

**Comment by mdhaber** [comment]
I went ahead and found the true inverse of the boxcox transform with respect to lambda so we could judge whether the approximation should be used instead.

```python3
from scipy import special, stats
import numpy as np

def boxcox_inv_lmbda(x, y):
  x = np.max(x)
  num = special.lambertw(- x**(-1/y) * np.log(x) / y, k=1)
  return np.real(-num/np.log(x) - 1/y)

def boxcox_inv_lmbda_approx(x, y):
  return np.log(y)/np.max(np.log(x))

x = np.array([2003.0, 1950.0, 1997.0, 2000.0, 2009.0])
ymax = np.finfo(x.dtype).max/100

lmbda = boxcox_inv_lmbda(x, ymax)
lmbda2 = boxcox_inv_lmbda_approx(x, ymax)
print(lmbda, lmbda2)
# 93.3171560299841 92.72073140759717

y = stats.boxcox(x, lmbda)
y2 = stats.boxcox(x, lmbda2)
print(np.max(y), np.max(y2), ymax)
# 1.7977637821054136e+306 1.9388254466627423e+304 1.7976931348623156e+306

np.testing.assert_allclose(np.max(y), ymax, rtol=1e-3)  # passs
```

It does seem like the approximation is conservative, though, so I guess it's OK. 

Can I tempt you to make the call @lucascolley? Quick overview:
- The purpose of `boxcox` is to transform data `x` into more normally distributed data `y`
- `boxcox_normmax` optimizes a parameter required by `boxcox: `lmbda`
- For some data, the optimal value of `lmbda` would cause `y` to overflow. 
- The purpose here is to find a value of `lmbda` that is close to the optimal value but doesn't making `y` overflow. 
- We do that by solving `np.finfo(x.dtype).max / safety_factor = stats.boxcox(x, lmbda)` for `lmbda`
- The question is whether to do it approximately or exactly.

**Comment by xuefeng-xu** [comment]
Thanks @mdhaber, this is accurate.

I came cross this topic from a paper[1], they found overflow occurs in Yeo-Johnson transformation in scikit-learn's implementation. Then I checked scikit-learn's issue and PR, which leads me to scipy. I think the overflow in Box-Cox transformation (positive data only) might be easier to solve than YJ (positive and negative data) based on prior work, thus I opened this PR.

[1] https://arxiv.org/abs/2210.01639

**Comment by lucascolley** [comment]
> I went ahead and found the true inverse of the boxcox transform with respect to lambda so we could judge whether the approximation should be used instead.

I read through some of the discussion in gh-18852. @xuefeng-xu if you are happy to switch to the true inverse approach, I think that would be best. Starting off with the best accuracy, and only sacrificing it if there is a use case which needs better performance seems like the right way to go. We can keep the approximation in the back of our mind (or perhaps add a comment into the code) in case we get such a performance request in the future.

On the other hand, it doesn't seem to make a huge difference, so if you would really rather stick with the approximation, I'm happy with that too.

**Comment by xuefeng-xu** [comment]
> if you are happy to switch to the true inverse approach, I think that would be best

Sure, I can switch to the true inverse.

But I would like to hear your opinions about the value of safety factor, maybe we can use 100?

**Comment by xuefeng-xu** [comment]
> ```python
> x = np.array([2003.0, 1950.0, 1997.0, 2000.0, 2009.0]) / 10
>
> 133.9054352555014 132.98192948755812 # lmbda
> inf 1.351832645074157e+304 1.7976931348623156e+306 # y
> ```

I saw overflow occurs when I divided x by 10 for the exact approach, maybe we should discuss about the safety factor.

**Comment by mdhaber** [comment]
Since `boxcox_normmax`'s job is to compute the optimal `lmbda`, and that can be done without consideration of whether the transformation itself will overflow, we discussed for Yeo-Johnson that the user should have the option to do so. A simple way to do that, which would also avoid the need to pick the one "best" safety factor for all uses, is to expose an option that controls the maximum value of the transformed data. I'd have to think about a name, but suppose it's called `ymax`, and it takes the place of `finfo(...).max/100`. The default could either be `inf` (return true optimum) or some value we think is safe. We could simply document that if the value exceeds the default, the true optimum `lmbda` will be returned. This would avoid the problem with overflow when the limit is set too high.  I think we could also fix the overflow problem, really, so let's not let that complicate the decision.

Maybe the extra argument in favor of the exact approach is that this would be relatively easy to document. If we were to use the approximation, and if we expose an option like this, then the meaning of the parameter would be more abstract to the user.

**Comment by xuefeng-xu** [comment]
I agree to provide an another parameter and then use the exact approach. Just to check my understanding.

> The default could either be inf (return true optimum) or some value we think is safe. We could simply document that if the value exceeds the default, the true optimum lmbda will be returned.

Do you mean return the constrained optimum when the value exceeds the default? (If `ymax` is not inf)

**Comment by mdhaber** [comment]
> Do you mean return the constrained optimum when the value exceeds the default?

Depends on what "the value" means.

Let's assume we can fix premature overflow issues and make it so the default `ymax` is the maximum float of the input data's dtype. Then by default, the function returns the optimal `lmbda_opt` subject to the constraint that the maximum of the data transformed by `boxcox` with parameter `lmbda_opt` does not overflow.

If the user sets `ymax=1e30`, then the function returns optimal `lmbda_opt` subject to the constraint that the maximum of the transformed data does not exceed `1e30`.

If the user sets `ymax=np.inf`, then the function returns the true, unconstrained optimal `lmbda_opt`, which could potentially cause the transformed data to overflow.

**Comment by mdhaber** [comment]
> I saw overflow occurs when I divided x by 10 for the exact approach, maybe we should discuss about the safety factor. 

This is because`boxcox` itself overflows prematurely. This *could* be fixed by trying to compute `boxcox` in log-space if it the usual calculation overflows, e.g.

```python3
def boxcox(x, lmbda):
  return np.exp(lmbda*np.log(x) - np.log(lmbda)) - 1/lmbda
```

It would probably be better to take care of that in `scipy.special`.

**Comment by xuefeng-xu** [comment]
I have updated code to use the exact approach. Since add `ymax` will change the API, I will later open another PR after applying the same idea here in Yeo-Johnson transformation.

**Comment by mdhaber** [comment]
> I will later open another PR after applying the same idea here in Yeo-Johnson transformation.

Sorry, I didn't understand. Do you mean you don't want to add `ymax` in this PR, or are you just saying you also want to help with YJ?

Either way, let's see if @lorentzenchr and @lsorber are happy with this plan. 
Christian, Laurent - the idea is to add a parameter `ymax` that allows the user to specify the maximum acceptable value of the transformed data.
For example, if the user sets `ymax=1e30`, then the function returns optimal `lmbda` subject to the constraint that the data transformed with `lmbda` does not exceed `1e30`.
 If they set `ymax=np.inf`, we return the true, unconstrained optimum `lmbda` (which we can determine without overflow by working in log space). We would do the same with the Yeo-Johnson transform after Box-Cox.

**Comment by xuefeng-xu** [comment]
Sorry for the confusion. I think maybe add `ymax` can be done in another PR, and I also want help with YJ.

**Comment by mdhaber** [comment]
@lucascolley how does [that](https://github.com/scipy/scipy/pull/19604#issuecomment-1835543246) sound to you? Any changes you'd make here?

**Comment by lorentzenchr** [comment]
@mdhaber `y_max` would give very good control for a user. On the other side, most of the time I, as a user, do not want to care about that. So at least something like `y_max=None` would be a nice default.

Another possibility is to raise a warning in case of overflow of transformed y and print the max possible lambda without overflow in the warning message.

**Review by lucascolley — APPROVED** [review]
> how does https://github.com/scipy/scipy/pull/19604#issuecomment-1835543246 sound to you? Any changes you'd make here?

If you are happy Matt then let's just get this improvement in and address giving the user more control in a follow up, keeping @lorentzenchr's comments in mind 👍

File: scipy/stats/_morestats.py [hunk 3]
```diff
@@ -1288,6 +1309,28 @@ def _all(x):
...
+        # Test if the optimal lambda causes overflow
+        x = np.asarray(x)
+        max_x = np.max(x, axis=0)
+        istransinf = np.isinf(special.boxcox(max_x, res))
+        dtype = x.dtype if np.issubdtype(x.dtype, np.floating) else np.float64
+        if np.any(istransinf):
+            warnings.warn(
+                f"The optimal lambda is {res}, but the returned lambda is "
+                f"the constrained optimum to ensure that maximum of "
+                f"transformed data does not cause overflow in {dtype}."
```
[the remark this answers is not in the document]
Comment (xuefeng-xu) [reply scipy/stats/_morestats.py]: done

**Comment by mdhaber** [comment]
CI failure is unrelated, and we've seen tests pass on Cirrus before. This LGTM, too, so let's merge and add the `ymax` or similar parameter in a followup.

**Comment by lorentzenchr** [comment]
@mdhaber If you want, you can tag me on that future PR, too.

**Comment by lucascolley** [comment]
> What do you think about turning on B028 in linting?

Happy to make a PR for this unless someone beats me to it

**Pull request merged**